pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -808,21 +808,31 @@ func (m *ManagerImpl) filterByAffinity(podUID, contName, resource string, availa
 	var withoutTopology []string
 	for d := range available {
 		// Since the same device may be associated with multiple NUMA Nodes. We
-		// need to be careful not to add each device to multiple lists. The
-		// logic below ensures this by breaking after the first NUMA node that
-		// has the device is encountered.
+		// need to be careful not to add each device to multiple lists. The logic
+		// below ensures this by loop throug all NUMA Nodes and select a target
+		// list from the three devices list defined above and then insert the device
+		// into this list. For NUMA node that has the device, this loop will break
+		// only when it found that this node meets the affinity requirement and
+		// set fromAffinity as target list. Otherwise it will keep looping and:
+		//    i: set nodeWithoutTopology as target list when targetList is nil and
+		//   ii: set notFromAffinity as target list when target is nil or nodeWithoutTopology
+		var targetList *[]string
```

Comment:
I'd prefer to avoid this indirection in the code

Reply:
@fromanirh I think a better way to fix this problem is to sort nodes list used in the loop in the order that nodes with numa affinity comes first then nodes without numa affinity and nodes without topology come as last, then we  no longer need the indirection list. If this is ok for you, I'll take another try.

Reply:
This seems a bit better, yes.